Mathlib/RingTheory/AdicCompletion/RingHom.lean

```diff
@@ -0,0 +1,282 @@
+/-
+Copyright (c) 2025 Jiedong Jiang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jiedong Jiang
+-/
+import Mathlib.RingTheory.AdicCompletion.Algebra
+
+/-!
+# Lift of ring homomorphisms to adic completions
+Let `R`, `S` be rings, `I` be an ideal of `S`.
+In this file we prove that a compatible family of ring homomorphisms from a ring `R` to
+`S ⧸ I ^ n` can be lifted to a ring homomorphism `R →+* AdicCompletion I S`.
+If `S` is `I`-adically complete, then this compatible family of ring homomorphisms can be
+lifted to a ring homomorphism `R →+* S`.
+
+## Main definitions
+
+- `AdicCompletion.liftRingHom`: given a compatible family of ring maps
+  `R →+* S ⧸ I ^ n`, the lift ring map `R →+* AdicCompletion I S`.
+- `IsAdicComplete.liftRingHom`: if `R` is
+  `I`-adically complete, then a compatible family of
+  ring maps `S →+* R ⧸ I ^ n` can be lifted to a unique ring map `S →+* R`.
+  Together with `mk_liftRingHom_apply` and `eq_liftRingHom`, it gives the universal property
+  of `R` being `I`-adically complete.
+-/
+
+open Ideal Quotient
+
+variable {R S : Type*} [NonAssocSemiring R] [CommRing S] (I : Ideal S)
+
+namespace AdicCompletion
+
+/--
+The composition map `S →+* AdicCompletion I S →+* S ⧸ I ^ n` equals to the natural quotient map.
+-/
+@[simp]
+theorem evalₐ_of_apply (n : ℕ) (x : S) :
+    (evalₐ I n) (of I S x) = Ideal.Quotient.mk _ x := by
+  simp [evalₐ, -smul_eq_mul, -Algebra.id.smul_eq_mul]
+
+theorem surjective_evalₐ (n : ℕ) : Function.Surjective (evalₐ I n) := by
+  simp only [evalₐ, smul_eq_mul, quotientEquivAlgOfEq_coe_eq_factorₐ,
+    AlgHom.coe_comp]
+  apply Function.Surjective.comp
+  · exact factor_surjective mul_le_right
+  · exact eval_surjective I S n
+
+/--
+The universal property of `IsAdicComplete` for rings.
+The lift ring map `R →+* AdicCompletion I S` of a compatible family of
+ring maps `R →+* S ⧸ I ^ n`.
+-/
+def liftRingHom (f : (n : ℕ) → R →+* S ⧸ I ^ n)
+    (hf : ∀ {m n : ℕ} (hle : m ≤ n), (factorPow I hle).comp (f n) = f m) :
+    R →+* AdicCompletion I S where
+  toFun := fun x ↦ ⟨fun n ↦ (factor (le_of_eq (mul_top _).symm)) (f n x),
+    fun hkl ↦ by simp [transitionMap, Submodule.factorPow, ← hf hkl]⟩
+  map_add' x y := by
+    simp only [map_add]
+    rfl
```

Comment:
Are there lemmas missing here?

Also I would prefer `by simp only [map_add]; rfl` because this proof is unimportant and shouldn't take up that much space but I think there aren't guidelines on this so this remark is non-blocking.